File: ios-backup/BodyEcho/Services/AnonymizationService.swift
//
//  AnonymizationService.swift
//  BodyEcho
//
//  Created on 2025-11-10
//

import Foundation
import CryptoKit

class AnonymizationService {
    static let shared = AnonymizationService()

    private init() {}

    /// Generates an anonymous user ID from Firebase UID
    func generateAnonymousId(from firebaseUid: String) -> String {
        let inputData = Data(firebaseUid.utf8)
        let hashed = SHA256.hash(data: inputData)
        return hashed.compactMap { String(format: "%02x", $0) }.joined()
    }

    /// Anonymizes email address
    func anonymizeEmail(_ email: String) -> String {
        let components = email.components(separatedBy: "@")
        guard components.count == 2 else { return email }

        let username = components[0]
        let domain = components[1]

        // Show first 2 characters and hash the rest
        let visiblePart = String(username.prefix(2))
        let hiddenPart = String(repeating: "*", count: max(username.count - 2, 3))

        return "\(visiblePart)\(hiddenPart)@\(domain)"
    }

    /// Encrypts sensitive data (simple implementation)
    func encryptData(_ data: String) -> String {
        let inputData = Data(data.utf8)
        let hashed = SHA256.hash(data: inputData)
        return hashed.compactMap { String(format: "%02x", $0) }.joined()
    }
}

File: lib/services/anonymization_service.dart
import 'dart:convert';
import 'package:crypto/crypto.dart';

class AnonymizationService {
  // Generate anonymous ID from Firebase UID
  String generateAnonymousId(String firebaseUid) {
    final bytes = utf8.encode(firebaseUid);
    final digest = sha256.convert(bytes);
    return digest.toString();
  }

  // Anonymize email address
  String anonymizeEmail(String email) {
    final parts = email.split('@');
    if (parts.length != 2) return email;

    final username = parts[0];
    final domain = parts[1];

    final visiblePart = username.length >= 2 ? username.substring(0, 2) : username;
    final hiddenPart = '*' * (username.length - 2).clamp(3, 10);

    return '$visiblePart$hiddenPart@$domain';
  }

  // Encrypt sensitive data
  String encryptData(String data) {
    final bytes = utf8.encode(data);
    final digest = sha256.convert(bytes);
    return digest.toString();
  }
}
